### PR TITLE
Roles and refactor

### PR DESCRIPTION
**Pull request by chelnak**

### CODE REVIEW

File: src/SFA.DAS.ToolService.Core/Configuration/RouteNames.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    }
+
+    public class AdminRoleRouteNames
+    {
+        public const string ManageRoles = "ManageRoles";
+        public const string SyncRoles = "SyncRoles";
+    }
+
+}
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Core/Configuration/RouteNames.cs]: I'm not sure if RouteNames should live in Core, they are frontend specific so should probably live in the Web project

File: src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using SFA.DAS.ToolService.Core.Entities;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.ToolService.Core.IRepositories
+{
+    public interface IAuth0RoleRepository
+    {
+        Task<List<Auth0Role>> GetRoles();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs]: If this is Auth0 specific then it should probably be GetAuth0Roles

File: src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs [hunk 3]
```diff
@@ -10,55 +11,63 @@ namespace SFA.DAS.ToolService.Core.Services
...
-        private readonly IApplicationRepository _applicationRepository;
+        private readonly ILogger<ApplicationService> logger;
+        private readonly IApplicationRepository applicationRepository;
 
-        public ApplicationService(IApplicationRepository applicationRepository)
+        public ApplicationService(ILogger<ApplicationService> _logger, IApplicationRepository _applicationRepository)
         {
-            _applicationRepository = applicationRepository;
+            logger = _logger;
+            applicationRepository = _applicationRepository;
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs]: Why's the `_` been inverted? Revert pls

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+using Microsoft.Extensions.Options;
+using SFA.DAS.ToolService.Core.Configuration;
+using Auth0.ManagementApi.Models;
+using Auth0.ManagementApi.Paging;
+using RestSharp;
+using Newtonsoft.Json;
+
+namespace SFA.DAS.ToolService.Infrastructure.Auth0
+{
+    public class ApiClient : IApiClient
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs]: Auth0ApiClient

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/IApiClient.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+
+using Auth0.ManagementApi.Paging;
+using Auth0.ManagementApi.Models;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.ToolService.Infrastructure.Auth0
+{
+    public interface IApiClient
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Auth0/IApiClient.cs]: IAuth0ApiClient

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs [hunk 6]
```diff
@@ -24,7 +24,7 @@ public async Task<List<Application>> GetApplicationsInRole(int id)
         {
             var result = await _toolServiceDbContext.ApplicationRole
                 .Where(c => c.RoleId.Equals(id) & c.Application.Public.Equals(0))
-                .Select(c => c.Application).ToListAsync();
+                .Select(c => c.Application).OrderBy(c => c.Name).ToListAsync();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: Ordering should probably be done higher up the stack, I think it's better that EntityFramework does as little as possible

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs [hunk 7]
```diff
@@ -37,30 +37,44 @@ public async Task<List<Application>> GetApplicationsNotInRole(int id)
                 .Select(c => c.ApplicationId).ToArrayAsync();
 
             var result = await _toolServiceDbContext.Application
-                .Where(c => !roleMappings.Contains(c.Id)).ToListAsync();
-           
+                .Where(c => !roleMappings.Contains(c.Id) & c.Public == 0)
+                .OrderBy(c => c.Name).ToListAsync();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: As above

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs [hunk 8]
```diff
@@ -37,30 +37,44 @@ public async Task<List<Application>> GetApplicationsNotInRole(int id)
...
         }
 
+        public async Task<List<Application>> GetApplicationsWithNoRoleAssignment()
+        {
+            var roleMappings = await _toolServiceDbContext.ApplicationRole
+                .Select(c => c.ApplicationId).ToArrayAsync();
+
+            var result = await _toolServiceDbContext.Application
+                .Where(c => !roleMappings.Contains(c.Id) & c.Public == 0)
+                .OrderBy(c => c.Name).ToListAsync();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: As above

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs [hunk 9]
```diff
@@ -37,30 +37,44 @@ public async Task<List<Application>> GetApplicationsNotInRole(int id)
...
+
+            return result;
+        }
         public async Task<List<Application>> GetPublicApplications()
         {
             var result = await _toolServiceDbContext.Application
-                .Where(c => c.Public.Equals(1)).ToListAsync();
+                .Where(c => c.Public.Equals(1))
+                .OrderBy(c => c.Name).ToListAsync();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ""

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs [hunk 10]
```diff
@@ -37,30 +37,44 @@ public async Task<List<Application>> GetApplicationsNotInRole(int id)
...
 
             return result;
         }
 
         public async Task<List<Application>> GetApplications()
         {
-            var result = await _toolServiceDbContext.Application.ToListAsync();
+            var result = await _toolServiceDbContext.Application
+                .OrderBy(c => c.Name).ToListAsync();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ""

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs [hunk 11]
```diff
@@ -37,30 +37,44 @@ public async Task<List<Application>> GetApplicationsNotInRole(int id)
...
 
             return result;
         }
 
         public async Task<List<Role>> GetRoles()
         {
             var result = await _toolServiceDbContext.Role
-                .ToListAsync();
+                .OrderBy(c => c.Name).ToListAsync();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ""

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/Auth0RoleRepository.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public class Auth0RoleRepository : IAuth0RoleRepository
+    {
+        private readonly ILogger<Auth0RoleRepository> logger;
+        private readonly IApiClient client;
+        public Auth0RoleRepository(ILogger<Auth0RoleRepository> _logger, IApiClient _client)
+        {
+            logger = _logger;
+            client = _client;
+        }
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/Auth0RoleRepository.cs]: Invereted `_`s again

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    {
+
+        private readonly ILogger<ApiClient> logger;
+        private readonly IOptions<AuthenticationConfiguration> configuration;
+
+        public ApiClient(ILogger<ApiClient> _logger, IOptions<AuthenticationConfiguration> _configuration)
+        {
+            logger = _logger;
+            configuration = _configuration;
+        }
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs]: Inverted `_`s

File: src/SFA.DAS.ToolService.Core/Services/RoleService.cs [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        private readonly ILogger<RoleService> logger;
+        private readonly IAuth0RoleRepository identityProviderRoleRepository;
+        private readonly IRoleRepository localRoleRepository;
+
+        public RoleService(ILogger<RoleService> _logger, IAuth0RoleRepository _identityProviderRoleRepository, IRoleRepository _localRoleRepository)
+        {
+            logger = _logger;
+            identityProviderRoleRepository = _identityProviderRoleRepository;
+            localRoleRepository = _localRoleRepository;
+        }
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Core/Services/RoleService.cs]: `_`

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/RoleRepository.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    public class RoleRepository : IRoleRepository
+    {
+        private readonly IToolServiceDbContext toolServiceDbContext;
+        private readonly ILogger<RoleRepository> logger;
+
+        public RoleRepository(IToolServiceDbContext _toolServiceDbContext, ILogger<RoleRepository> _logger)
+        {
+            toolServiceDbContext = _toolServiceDbContext;
+            logger = _logger;
+        }
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Repositories/RoleRepository.cs]: `_`

File: src/SFA.DAS.ToolService.Core/Configuration/RouteNames.cs (hunk 1 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Core/Configuration/RouteNames.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs (hunk 4 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/IApiClient.cs (hunk 5 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Auth0/IApiClient.cs]: ✔️

File: src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs (hunk 2 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs (hunk 3 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/Auth0RoleRepository.cs (hunk 12 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/Auth0RoleRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs (hunk 13 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Auth0/ApiClient.cs]: ✔️

File: src/SFA.DAS.ToolService.Core/Services/RoleService.cs (hunk 14 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Core/Services/RoleService.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/RoleRepository.cs (hunk 15 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/RoleRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs (hunk 6 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs (hunk 7 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs (hunk 8 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs (hunk 9 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs (hunk 10 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs (hunk 11 above)
Comment (chelnak) [reply src/SFA.DAS.ToolService.Infrastructure/Repositories/ApplicationRepository.cs]: ✔️

File: src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs (hunk 3 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs]: ???

File: src/SFA.DAS.ToolService.Core/Configuration/RouteNames.cs (hunk 1 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/Configuration/RouteNames.cs]: ???

File: src/SFA.DAS.ToolService.Core/Services/RoleService.cs (hunk 14 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/Services/RoleService.cs]: ???

File: src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs (hunk 2 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs]: ???

File: src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs (hunk 3 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/Services/ApplicationService.cs]: ???

File: src/SFA.DAS.ToolService.Core/Services/RoleService.cs (hunk 14 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/Services/RoleService.cs]: ???

File: src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs (hunk 2 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/IRepositories/IAuth0RoleRepository.cs]: Lets change the interface to something like IExternalRoleRepository

File: src/SFA.DAS.ToolService.Core/Services/RoleService.cs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+        public async Task SyncIdentityProviderRoles()
+        {
+            // TODO: try harder
+            var idpRoles = await _identityProviderRoleRepository.GetRoles();
+
+            foreach (var role in idpRoles)
+            {
+                var dbEntity = await _localRoleRepository.GetRole(role.Name);
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Core/Services/RoleService.cs]: Wouldn't the ID be more suitable as a lookup property than the name string?

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/Auth0ApiClient.cs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+        private readonly ILogger<Auth0ApiClient> _logger;
+        private readonly IOptions<AuthenticationConfiguration> _configuration;
+
+        public Auth0ApiClient(ILogger<Auth0ApiClient> logger, IOptions<AuthenticationConfiguration> configuration)
+        {
+            _logger = logger;
+            _configuration = configuration;
+        }
+
+        public class TokenResponse
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Auth0/Auth0ApiClient.cs]: private?

File: src/SFA.DAS.ToolService.Core/IRepositories/IExternalRoleRepository.cs [hunk 18]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Collections.Generic;
+using SFA.DAS.ToolService.Core.Entities;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.ToolService.Core.IRepositories
+{
+    public interface IExternalRoleRepository
+    {
+        Task<List<Auth0Role>> GetExternalRoles();
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Core/IRepositories/IExternalRoleRepository.cs]: `Auth0Role` now should probably inherit from a generic `ExternalRole` class to fully abstract this repository

File: src/SFA.DAS.ToolService.Core/Services/RoleService.cs (hunk 16 above)
Comment (EwanNoble) [reply src/SFA.DAS.ToolService.Core/Services/RoleService.cs]: Issue created

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/Auth0ApiClient.cs [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+            public long ExpiresIn { get; set; }
+            [JsonProperty("scope")]
+            public string Scope { get; set; }
+            [JsonProperty("token_type")]
+            public string TokenType { get; set; }
+        }
+
+    private string GetToken()
+        {
+            var client = new RestClient($"https://{_configuration.Value.Domain}/oauth/token");
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Auth0/Auth0ApiClient.cs]: Can we put `using ()` here to ensure the client is disposed of?

File: src/SFA.DAS.ToolService.Infrastructure/Auth0/Auth0ApiClient.cs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+
+            var tokenResponse = JsonConvert.DeserializeObject<TokenResponse>(response.Content);
+            return tokenResponse.AccessToken;
+        }
+
+        public async Task<IPagedList<Role>> GetAuth0Roles()
+        {
+            var token = GetToken();
+            var uri = new Uri($"https://{_configuration.Value.Domain}/api/v2");
+            var client = new ManagementApiClient(token, uri);
```
Comment (EwanNoble) [inline src/SFA.DAS.ToolService.Infrastructure/Auth0/Auth0ApiClient.cs]: `using ()` possible here?

**Pull request merged**